File: cs50x/Week 01/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main()
{
    int sum = 0, digits = 0;
    long number, starts;
    number = starts = get_long("Number: ");

    while (starts >= 100)
    {
        starts /= 10;
    }

    while (number)
    {
        // last number
        int product = number % 10;
        sum += product;

        number = number / 10;
        digits++;

        // second last
        if (number)
        {
            product = number % 10 * 2;

            if (product >= 10)
            {
                sum += product % 10;
                product /= 10;
            }

            sum += product;

            number = number / 10;
            digits++;
        }
    }

    if (sum % 10 == 0)
    {
        if (digits == 15 && (starts == 34 || starts == 37))
        {
            printf("AMEX\n");
        }
        else if (digits == 16 && (starts >= 51 && starts <= 55))
        {
            printf("MASTERCARD\n");
        }
        else if ((digits == 13 || digits == 16) && (starts >= 40 && starts <= 49))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}

File: cs50x/Week 06/credit/credit.py
import re


def main():
    print(validate(int(input("Number: "))))


def validate(number):
    str_copy = str(number)

    sum = 0

    while number:
        last = number % 10
        number = int(number / 10)

        sum += last

        last = number % 10 * 2
        number = int(number / 10)

        sum += last // 10 + last % 10

    if sum % 10 != 0:
        return "INVALID"

    number = str_copy

    amex = r"^(?:34|37)\d{13}"
    mastercard = r"^(?:51|52|53|54|55)\d{14}"
    visa = r"^4(?:\d{12}|\d{15})"

    if re.match(amex, number):
        return "AMEX"
    elif re.match(mastercard, number):
        return "MASTERCARD"
    elif re.match(visa, number):
        return "VISA"
    else:
        return "INVALID"


if __name__ == "__main__":
    main()
